search.py

```diff
@@ -484,6 +484,130 @@ def h(self, node):
+
+        if index_blank_square % self.size == 0:
+            possible_actions.remove('LEFT')
+        if index_blank_square < self.size:
+            possible_actions.remove('UP')
+        if index_blank_square % self.size == self.size -1 :
+            possible_actions.remove('RIGHT')
+        if index_blank_square > self.size*self.size - self.size - 1:
+            possible_actions.remove('DOWN')
+
+        return possible_actions
+
+    def result(self, state, action):
+        """ Given state and action, return a new state that is the result of the action.
+        Action is assumed to be a valid action in the state """
+
+        index_blank_square = self.find_blank_square(state)
+        new_state = list(state)
+
+        delta = {'UP': -self.size, 'DOWN': self.size, 'LEFT': -1, 'RIGHT': 1}
+        neighbor = index_blank_square + delta[action]
+        new_state[index_blank_square], new_state[neighbor] = new_state[neighbor], new_state[index_blank_square]
+
+        return tuple(new_state)
+
+    def goal_test(self, state):
+        """ Given a state, return True if state is a goal state or False otherwise """
+
+        return state == self.goal
+
+    def check_solvability(self, state):
+        """ Checks if the given state is solvable. Whether or not puzzle is solvable depends on size being even or odd.
+                     If N is odd, then puzzle instance is solvable if number of inversions is even in the input state.
+                    If N is even, puzzle instance is solvable if the blank is on an even row counting from the bottom (second-last, fourth-last, etc.) and number of inversions is odd
+                    or the blank is on an odd row counting from the bottom (last, third-last, fifth-last, etc.) and number of inversions is even."
+                    reference : https://www.geeksforgeeks.org/check-instance-15-puzzle-solvable/"""
+
+        solvable = True
+        inversions = 0
+        for i in range(len(state)-1):
+            for j in range(i+1, len(state)):
+                if state[i] != 0 and state[j] != 0 and state[i] > state[j]:
+                    inversions += 1
+
+        if self.size % 2 != 0:
+            solvable = True if (inversions % 2 == 0 ) else False
+        elif self.find_blank_square(state) // self.size % 2 == 0 and inversions % 2 != 0 :
+            solvable = True
+        elif self.find_blank_square(state) // self.size % 2 != 0 and inversions % 2 == 0:
+            solvable = True
+        else:
+            solvable = False
+
+
+        return solvable
+
+    def h(self, node):
+        """ Return the heuristic value for a given state. Default heuristic function used is
+        h(n) = number of misplaced tiles """
+        heuristics = {
+            'manhattan_distance':self.manhattan_distance_heuristic(node),
+            'misplaced_tiles':self.misplaced_tiles_heuristic(node)
+        }
+        return heuristics.get(self.heuristic)
+
+    def manhattan_distance_heuristic(self, node):
```

Comment:
I'm not intimately familiar with the repo and the way the maintainers like things, but it may be worth calling the `manhattan_distance` and `hamming_distance` methods from [learning.py](https://github.com/aimacode/aima-python/blob/c16bb8bdc28e8f9fcc6e7f76a92b9492bf019d87/learning.py) to avoid reimplementing code already in the repo.

Reply:
Good catch!

I would go even further and say that the distance functions should be moved to `utils.py`, since they are not directly tied to learning algorithms.

Reply:
Thank you for the suggestion! :)  I'm not familiar with the contribution guidelines as well. So I removed `manhattan_distance`  (the implementation in learning.py is applicable only to 1 dimension distance )  and used `hamming_distance` from `learning.py` isntead.